### ISSUE TITLE
Consider renaming local_datetime back to naive_datetime

### ISSUE THREAD

**Issue by 1st1** [edgedb/edgedb#866]
A few questions regarding our handling of date/time came up in our internal Slack recently:

* I would like UTC timestamp without any timezone; should use local_datetime? 
* Or should I use datetime and expect nobody messes up with timezone? 
* If I need a timestamp, like expiration time of something. Should this be a timezone-aware datetime? 
* Should I assume it's UTC when returned from database or should I enforce (convert) timezone when formatting?
* I would expect `local_datetime` to be timezone-aware.

Since this is quite a complex topic I'd like to start by explaining how EdgeDB works with date/time compared to Postgres.

### EdgeDB

In EdgeDB there is a timezone aware type `std::datetime` and a naive date/time type `std::local_datetime`.

`std::datetime` is stored internally in UTC and binary-wise it's identical to `std::local_datetime`. The fact that `std::datetime` values are time zone aware is **encoded into their type**.  E.g. when a user receives a `std::datetime` value they:

* Know that they are dealing with a time zone aware date/time.

* If they want to render this date time to string in their front-end code they have to specify the target time zone.  E.g. to render a `std::datetime` value to a string via EdgeQL the user would use `std::to_str` function that would require the user to pass the timezone they want the time to be rendered in. Usually that timezone would be the timezone of the current user that's interacting with the date.

* EdgeQL is strict w.r.t. separating `std::datetime` from `std::local_datetime`. For instance it's not possible to subtract `std::datetime` from `std::local_datetime` — the operation is undefined. Subtracting `std::datetime` from `std::datetime` or `std::local_datetime` from `std::local_datetime` is allowed and would produce a `std::duration`.

* When a user submits a `std::datetime` value to the database it's their responsibility to let the DB know how to convert it to UTC.  There's no magical config setting (unlike in Postgres) that would automatically let naive date/time be converted to time zone aware date/time.

**Here's a couple of use cases to understand the naming motivation:**

* `datetime` is a globally recognizable point in time. Example: space shuttle launch time; file creation time; when a log or audit record was created; precise moment of when a financial transaction has been committed.

* `local_datetime` is something that makes sense regardless of time zone. Examples: milk expiry date; alarm clock wake up time; arrival time on air/train tickets is always in *local time of the destination*.

💡 Maybe we should add this ^^ to the docs.

### PostgreSQL

Quoting Postgres [documentation](https://www.postgresql.org/docs/12/datatype-datetime.html):

> PostgreSQL endeavors to be compatible with the SQL standard definitions for typical usage. However, the SQL standard has an odd mix of date and time types and capabilities. 

There are two types in Postgres that are interesting for us in the context of this discussion: `timestamp` and `timestamptz` (it's an alias for SQL's `timestamp with time zone`):

* `timestamp` is used to store naive date/time, e.g. when time zones aren't applicable or not known.

* `timestamptz` is used to store time zone aware date/time. It is stored on the server in UTC. As in EdgeDB, in Postgres, they are **identical** values when stored on the disk, meaning that the difference between the two is purely their *type id* (OID).

So far so good, Postgres date/time handling looks just like EdgeDB.  The key difference is that in Postgres there's a [server config setting](https://www.postgresql.org/docs/12/runtime-config-client.html#GUC-TIMEZONE) `TimeZone` that is used to **automatically convert** time zone aware and naive date/time to each other.  Examples:

* if a user submits a text date/time without timezone to `timezonetz` cast Postgres will use the config setting and convert it silently. When the user queries the submitted value back, it might be different from what was submitted because of that transformation, leaving the user confused.

* if a user subtracts a `timezone` value from a `timezonetz` value, Postgres will use the config setting to silently convert `timezone` to `timezonetz`. Again, this can lead to confusion, as the result would look totally different from what a naive user assumed.

**EdgeDB does not have automatic conversion.** To submit a `std::datetime` value a time zone must be specified along with it. *Implicit mixing* of time zone aware and naive date/time values is **prohibited**.

### Q&A

> * I would like UTC timestamp without any timezone; should use local_datetime? 

You should use `std::datetime`. Just point out to EdgeDB that it's already in UTC.

> * Or should I use datetime and expect nobody messes up with timezone? 

You should indeed use `std::datetime`. You should make sure that your frontend code renders your date/time values either in: local user's time zone; or in UTC. The format is up to you and your UI requirements, EdgeDB does not care.

> * If I need a timestamp, like expiration time of something. Should this be a timezone-aware datetime? 

Depends. If it's the expiration time of a carton of milk I'd say you should use `std::local_datetime`.  If it's the expiration time of evaluation period of your SaaS offering it should be `std::datetime` -- that way the user will know the precise time no matter where they *currently* are, in London or in New York.

> * Should I assume it's UTC when returned from database or should I enforce (convert) timezone when formatting?

For `std::datetime` — yes!

For `std::local_datetime` — it can be whatever, the database does not know. Only the developer knows and can use it correspondingly.

**The rule of thumb here** is that if the date/time value is meant to be rendered using different time/zones it should always be `std::datetime`.

> * I would expect `local_datetime` to be timezone-aware.

Yes, I can see where the confusion is coming from. See the💡-paragraph section above for our motivation of naming it that way.

------

We might want to consider renaming `local_datetime` back to `naive_datetime`. **At the very least our documentation must be improved.**

**Comment by 1st1** [comment]
cc @edgedb/engineering

**Comment by tailhook** [comment]
For context:
I brought this topic to the internal chat because I got used to using naive datetimes in UTC in postgres/mysql for timestamps. This is probably due to the fact that postgres' datetime representation:
```
2018-05-07T15:01:22+00
```
... is usually a bit harder to parse as far as I know. I.e. python's `fromisoformat` requires `+02:00` format and `strptime`/`strftime` don't even have a timezone specifier. Same thing with formatting. Similar thing in javascript. And Rust is close to that.

------

Technically it's not relevant to the edgedb as its own driver should read timestamps in whatever format needed and return language's native datetime object/datatype. So adding pieces of documentation would help.

We might also bikeshed on a name a little bit. My favorite option would be (1):
* `datetime` -- combination of date and time, naive
* `timestamp` -- point in time, unix timestamp, represented with timezone if stringified

Except it might be confusing for people already got used to postgres types. Other options:
* (2) `datetime` vs `instant` (downside: instant is monotonic clock in rust, but I don't think it can ever be written into a DB)
* (3) `naive_datetime` vs `datetime` as mentioned above (by the way, "naive" term is not very python-specific, it is also [used in rust](https://docs.rs/chrono/0.4.9/chrono/))

Any other ideas?

-----

And a related question, should we represent timezone-aware timestamps in ISO format by default:
```
2018-05-07T15:01:22Z
```
instead of `+00`. That makes it clear that UTC is always used. (not sure about efficiency?)

**Comment by elprans** [comment]
@tailhook We don't represent timestamps as strings (*), all data is encoded in binary formats: https://edgedb.com/docs/internals/protocol/dataformats#std-datetime

(*) Except in  JSON mode, and there it's a proper ISO8601 string:

```
elvis> SELECT datetime_current();
["2019-11-24T18:44:41.190056+00:00"]
```

**Comment by tailhook** [comment]
> @tailhook We don't represent timestamps as strings (*), all data is encoded in binary formats:

Sure. I mean what you show in examples, what `to_str` does by default, what casting to string does (if it works) and similar cases.

**Comment by elprans** [comment]
`to_str` and cast also look OK to me:

```
elvis> SELECT <str>datetime_current();
{'2019-11-24T19:55:23.309872+00:00'}
elvis> SELECT to_str(datetime_current());
{'2019-11-24T19:55:32.180566+00:00'}
```

**Comment by tailhook** [comment]
> to_str and cast also look OK to me:
>```
>elvis> SELECT <str>datetime_current();
>{'2019-11-24T19:55:23.309872+00:00'}
>```

This is specifically my point: `+00:00` questions: in what conditions it can be non-zero? Should I tweak timezone? The `Z` is more like: here is a standard representation of datetime. It's also still unambiguous and ISO-compatible.

**Comment by elprans** [comment]
> in what conditions it can be non-zero?

Right now, EdgeDB always runs with UTC as the configured timezone, so the offset is always zero. We might change this stance, and allow configuring the server timezone, in which case the timezone will be that of the server.

> The Z is more like: here is a standard representation of datetime. It's also still unambiguous and ISO-compatible.

And so is [00:00](https://en.wikipedia.org/wiki/ISO_8601#Time_offsets_from_UTC).  Specifically, in this case we use the output of PostgreSQL's formatting function.  Trying to massage its output has a performance impact.

**Comment by 1st1** [comment]
@elprans
> Right now, EdgeDB always runs with UTC as the configured timezone, so the offset is always zero. We might change this stance, and allow configuring the server timezone, in which case the timezone will be that of the server.

I think "server timezone" is a thing of the past. What's "server timezone" in the context of modern deployment? What's "server"?  Today your application is hosted in an east coast datacenter, tomorrow it's on west cost, in a year it might be deployed in multiple data centers.

We should stick with having no configurable server timezone and gently guide our users to a better way of working with date/time.

@elprans
> And so is +00:00. Specifically, in this case we use the output of PostgreSQL's formatting function. Trying to massage its output has a performance impact.

We can have a simple wrapper function to change "+00:00" to "Z", the cost of that should be negligible. Not sure it's needed though.

@tailhook 
> We might also bikeshed on a name a little bit. My favorite option would be (1):
>
> datetime -- combination of date and time, naive
> timestamp -- point in time, unix timestamp, represented with timezone if stringified

Well, `datetime` is a time zone aware thing in most languages, even in Rust: https://docs.rs/chrono/0.4.0/chrono/struct.DateTime.html. I think using your terminology wouldn't make things clearer.

----

\<offtopic\>
It seems that Rust's `chrono::DateTime` stores the original timezone info along with the value. Sadly it's not possible to do that in EdgeDB as Postgres' `timezonetz` is just a UTC timestamp. So we're forced to make this choice too — our `datetime` is just a timestamp in UTC and our `local_datetime` is just a timestamp.
\</offtopic\>

----

Your point that some people might misinterpret what "local" means is valid though. I see a few options here:

1. Change nothing. We can address the potential confusion with better docs.

2. Rename "local_" to "naive_". We would have then `datetime`, `naive_datetime`, `naive_date`, and `naive_time` — looks pretty obvious to me. The only caveat is to explain that `datetime` is always in UTC (this is what @tailhook also proposes as his option (3) in https://github.com/edgedb/edgedb/issues/866#issuecomment-557914283).

3. Rename "local_" to "naive_". Rename `datetime` to `utc_datetime`.  We would have then `utc_datetime`, `naive_datetime`, `naive_date`, and `naive_time`.  Maybe this would be the most obvious option (albeit a bit verbose for my taste).

**Comment by tailhook** [comment]
> Well, datetime is a time zone aware thing in most languages, even in Rust: 

That's not the type used for timestamps. [`SystemTime`](https://doc.rust-lang.org/std/time/struct.SystemTime.html) is. Chrono is only needed if you want complex date calculations (i.e. same day next month), or timezone awareness. I.e. working with real dates, not point in time.

I mean things like log timestamps, cache expiration times, etc. all work with SystemTime.

**Comment by 1st1** [comment]
> I mean things like log timestamps, cache expiration times, etc. all work with SystemTime.

Got it.  

What do you think about renaming `datetime` to `utc_datetime`?

**Comment by elprans** [comment]
> What do you think about renaming datetime to utc_datetime?

-1.  You can still have datetime values in whatever non-UTC timezone, and the server would happily accept them as input.  Also, it is conceivable to have a `to_str` variant that accepts a timezone, a-la `astimezone`.

**Comment by 1st1** [comment]
> -1. You can still have datetime values in whatever non-UTC timezone, and the server would happily accept them as input.

The server will accept them, but it **will convert them to UTC and store them as UTC**. Highlighting that is the point of the proposed name. I'm sorry, but you need to elaborate a bit more here than just -1-ning.

> Also, it is conceivable to have a to_str variant that accepts a timezone, a-la astimezone.

What do you mean? If a user wants to format `std::datetime` to text with a specific timezone they can do that. We also have functions to convert `local_datetime` to `datetime` with a specific timezone. Please clarify your point.

**Comment by elprans** [comment]
OK, let's take a step back and look at this from the user's perspective.  Most users will interact with EdgeDB through a driver, so the values they will send and receive will be instances of some timezone-aware language type, such as `datetime` in Python.  The driver may even convert the fetched `datetime` values to have a specific timezone.  The fact that the server is storing the value as a UTC timestamp is, IMO, not as relevant, compared to the overall mental model of the type as it integrates into the host language.

**Comment by elprans** [comment]
> I think "server timezone" is a thing of the past. What's "server timezone" in the context of modern deployment

The timezone setting does not have to be global, it may be per-session.

**Comment by tailhook** [comment]
Just discovered again: `datetime.utcnow()` is also naive. That's was also a tiny piece of reasoning I have always preferred storing a naive timestamp.

----

If we have `utc_datetime` and `naive_datetime` it's still not clear what to use for timestamps if judging only by name. It's easier to use naive in python at least because `utcnow()` works. That's why I prefer `timestamp`.

I also see `local_` or `naive_` prefixes as non-informative for `local_date` and `local_time`. This is also a reason why I like datetime being bare combination of date and time.

`utc_datetime` and `local_datetime`: makes me wonder what should I use for things like appointments? At a glance, it looks like I should use `local_datetime`, except I shouldn't (`naive_` might be better here).

These are just thoughts, not strong opinions.

> The driver may even convert the fetched datetime values to have a specific timezone

This is also something I questioned at first. But if it is like that. I would rather expect `local_datetime` to have this functionality from the newcomer's point of view.

**Comment by 1st1** [comment]
@elprans 
> The fact that the server is storing the value as a UTC timestamp is, IMO, not as relevant, compared to the overall mental model of the type as it integrates into the host language.

Alright, I see your point now, but I don't 100% agree. For example we have the `intN` types in EdgeDB but in Python all ints are of unlimited size. You can get away for a while not knowing about EdgeDB types when interfacing with it from Python, but at some point you might get an exception that your Python int value does not fit. At that point knowing what `int32` means in EdgeDB would help the user, even if their Python mental model is bended slightly.

There's also the mental model of EdgeQL users. And all users who will use EdgeQL in query builders. Our date/time functions and operators are designed with an assumption that the user understands our types.  If the user thinks that we're storing the original time zone in our `std::datetime` they would be confused and our API wouldn't make a lot of sense to them. Understanding how `std::datetime` is stored is crucial for the correct usage of our system, and a name like `std::utc_datetime` would certainly help.

Lastly, EdgeDB is designed to be used from multiple different languages. In a team that uses Go & Python there are at least three different mental models (Go, Python, EdgeQL). One of those mental models will be shared, and it's EdgeQL. So maybe we should focus on our main mental model for making this decision.

@elprans 
> The timezone setting does not have to be global, it may be per-session.

It's not clear to me — a timezone setting for what?  Default timezone for `to_str()` / other formatting?  Might be useful, but it's not related to the current naming question.  We won't add this before 1.0 anyways, the API churn has to stop. But if you're serious about it you should check if our current formatting functions would work with such setting. Ideally we want to avoid having many slightly different functions for formatting.

Or did you mean a setting for auto-converting naive to aware date/time though? This would be a big step back and kind of make our entire date/time design useless. We could just expose Postgres semantics directly if you wanted that.

All in all I'm not quite sure what you wanted this per-session setting to do...

@tailhook 
> I also see local_ or naive_ prefixes as non-informative for local_date and local_time. This is also a reason why I like datetime being bare combination of date and time.

Why aren't they informative? FWIW time zone isn't applicable to bare time or to bare date values, so they are always naive.

@tailhook 
> Just discovered again: datetime.utcnow() is also naive. That's was also a tiny piece of reasoning I have always preferred storing a naive timestamp.

Yes, that's an unfortunate design. Python documentation acknowledges it was a mistake to return a naive datetime for current UTC time; quoting the [big red warning](https://docs.python.org/3/library/datetime.html#datetime.datetime.utcnow) for this function:

> **Warning** Because naive datetime objects are treated by many datetime methods as local times, it is preferred to use aware datetimes to represent times in UTC. As such, the recommended way to create an object representing the current time in UTC by calling datetime.now(timezone.utc).

**Comment by elprans** [comment]
> It's not clear to me — a timezone setting for what?

Default timezone for string conversion, including in JSON.  I'm pretty sure this would work already if we exposed `TimeZone` in `cfg::Config`.  You can then `CONFIGURE SESSION` based on the user's preference and all dates automatically show in the desired time zone.

**Comment by 1st1** [comment]
After an intense discussion of this topic with @edgedb/engineering we decided to keep the status quo w.r.t. to naming. We do want to make sure that our documentation is thorough and explains the difference between `std::datetime` and `std::local_datetime` in great detail.

**Comment by tailhook** [comment]
[iCalendar uses](https://tools.ietf.org/html/rfc5545) name "floating date-time" so this is also a possible option for `local_datetime`. This is the only place I've seen it though.

**Comment by tailhook** [comment]
Decided to move it into a namespace (#902)

**Comment by buzinas** [comment]
> After an intense discussion of this topic with @edgedb/engineering we decided to keep the status quo w.r.t. to naming. We do want to make sure that our documentation is thorough and explains the difference between `std::datetime` and `std::local_datetime` in great detail.

Just wanted to point out that I looked throughout the entire documentation and I only understood / fixed my date types after reading this issue. Is there anything we can do to improve the docs on this?

**Comment by raddevon** [comment]
I'd like to improve documentation around this if there's an opportunity for it.

Here's our current documentation on `std::datetime`: https://www.edgedb.com/docs/stdlib/datetime#type::std::datetime

and here's what we have on `cal::local_datetime`: https://www.edgedb.com/docs/stdlib/datetime#type::cal::local_datetime

Would you mind sharing, @buzinas, where the gap was for you? I will try to figure out the best way to close that gap. Open to any suggestions you have on how to approach this too.

**Comment by buzinas** [comment]
@raddevon Thanks for the quick response!

Overall, the docs explain the difference in a very technical way (and not super intuitive, at least for me). Eg: what is a "timezone-aware" date in practice?

First I created my `createdAt` fields as `local_datetime` by converting current date to UTC. Only then I saw that datetime was automatically saved as UTC.

The use cases and the Q/A section of the original poster of this issue solved all the doubts I had about the two date types. At least part of that could be ported to the docs.
